dubbo-rpc/dubbo-rpc-dubbo/src/main/java/org/apache/dubbo/rpc/protocol/dubbo/DubboProtocol.java

```diff
@@ -396,12 +396,56 @@ private void optimizeSerialization(URL url) throws RpcException {
         optimizeSerialization(url);
 
         // create rpc invoker.
-        DubboInvoker<T> invoker = new DubboInvoker<T>(serviceType, url, getClients(url), invokers);
+        DubboInvoker<T> invoker = new DubboInvoker<T>(serviceType, url, getClientsV2(url), invokers);
```

Comment:
After you refactored your code, your unit tests failed. Please fix these exceptions.

[ERROR] DubboLazyConnectTest.testSticky1:53 Unexpected exception type thrown ==> expected: <org.apache.dubbo.rpc.RpcException> but was: <java.lang.IndexOutOfBoundsException>
[ERROR] DubboLazyConnectTest.testSticky3:67 Unexpected exception type thrown ==> expected: <org.apache.dubbo.rpc.RpcException> but was: <java.lang.IndexOutOfBoundsException>
[ERROR] ExplicitCallbackTest.TestCallbackConsumerLimit:176 Unexpected exception type thrown ==> expected: <org.apache.dubbo.rpc.RpcException> but was: <java.lang.IndexOutOfBoundsException>
[ERROR] ExplicitCallbackTest.TestCallbackProviderLimit:203 Unexpected exception type thrown ==> expected: <org.apache.dubbo.rpc.RpcException> but was: <java.lang.IndexOutOfBoundsException>
[ERROR] Errors:
[ERROR] DubboInvokerAvilableTest.test_Lazy_ChannelReadOnly:131 » IndexOutOfBounds Inde...
[ERROR] DubboInvokerAvilableTest.test_NoInvokers:118 » IndexOutOfBounds Index: 0, Size...
[ERROR] DubboInvokerAvilableTest.test_Normal_ChannelReadOnly:78 » IndexOutOfBounds Ind...
[ERROR] DubboInvokerAvilableTest.test_Normal_available:67 » IndexOutOfBounds Index: 0,...
[ERROR] DubboLazyConnectTest.testSticky2:62 » IndexOutOfBounds Index: 0, Size: 0
[ERROR] DubboLazyConnectTest.testSticky4:81 » IndexOutOfBounds Index: 0, Size: 0
[ERROR] DubboProtocolTest.testDemoProtocol:62 » IndexOutOfBounds Index: 0, Size: 0
[ERROR] DubboProtocolTest.testDubboProtocol:71 » IndexOutOfBounds Index: 0, Size: 0
[ERROR] DubboProtocolTest.testDubboProtocolMultiService:148 » IndexOutOfBounds Index: ...
[ERROR] DubboProtocolTest.testDubboProtocolWithMina:104 » IndexOutOfBounds Index: 0, S...
[ERROR] DubboProtocolTest.testNonSerializedParameter:178 » IndexOutOfBounds Index: 0, ...
[ERROR] DubboProtocolTest.testPerm:166 » IndexOutOfBounds Index: 0, Size: 0
[ERROR] DubboProtocolTest.testRemoteApplicationName:208 » IndexOutOfBounds Index: 0, S...
[ERROR] DubboProtocolTest.testReturnNonSerialized:192 » IndexOutOfBounds Index: 0, Siz...
[ERROR] ExplicitCallbackTest.TestCallbackMultiInstans:128->initOrResetService:90->referService:65 » IndexOutOfBounds
[ERROR] ExplicitCallbackTest.TestCallbackNormal:107->initOrResetService:90->referService:65 » IndexOutOfBounds
[ERROR] MultiThreadTest.testDubboMultiThreadInvoke:51 » IndexOutOfBounds Index: 0, Siz...
[ERROR] ReferenceCountExchangeClientTest.test_counter_error:156->init:223 » Arithmetic
[ERROR] ReferenceCountExchangeClientTest.test_mult_share_connect:117->init:223 » Arithmetic
[ERROR] ReferenceCountExchangeClientTest.test_multi_destory:139->init:223 » Arithmetic
[ERROR] ReferenceCountExchangeClientTest.test_not_share_connect:103->init:221->referInvoker:72 » IndexOutOfBounds
[ERROR] ReferenceCountExchangeClientTest.test_share_connect:92->init:223 » Arithmetic ...
[ERROR] RpcFilterTest.testRpcFilter:48 » IndexOutOfBounds Index: 0, Size: 0

@haiyang1985

Reply:
Current PR is for discuss only, not my final version.